Accept registration payments through Stripe - Fixes #729

pingskills opened the pull request:
Using Stripe Connect, a competition organiser can connect their own
Stripe account to the WCA Stripe account. Once this is done charges can
be made on the connected account and an application fee can be charged
by the WCA.

timhabermaas commented on WcaOnRails/app/controllers/registrations_controller.rb:
```diff
@@ -192,6 +192,47 @@ def register
+    end
+
+    registration.record_payment(
+      charge.amount,
```
`charge.amount` = `registration.outstanding_entry_fees.cents` ? If so, why not `registration.charge`?

pingskills replied:
`charge` is what stripe returns so although it will always be equal to `registration.outstanding_entry_fees.cents`, I thought it a good idea to record what Stripe sent back.

pingskills commented on WcaOnRails/app/controllers/registrations_controller.rb:
```diff
@@ -192,6 +192,47 @@ def register
+        source: token,
+        description: "Registration payment",
+        metadata: registration.user.wca_id,
+        application_fee: 100,
```
I probably should remove the application_fee for now, but this demonstrates how the WCA could take a fee for each registration. This would go straight into the WCA Stripe account.

larspetrus commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -262,6 +262,64 @@ def edit
+      state: @competition.id,
+    }
+    @authorise_url = client.auth_code.authorize_url(oauth_params)
+  end
```
Would this be too compact?

    @authorise_url = create_oauth_client.auth_code.authorize_url(scope: 'read_write', params[:id])

jfly replied:
It's probably worth noting that this rewrite may have an injection vulnerability thanks to using `params[:id]`. I feel much more comfortable doing a database query to find the competition first.

larspetrus replied:
OK, I didn't think of the param could be "unsafe". But loading the object from the DB is an awkward way of sanitizing it. It also makes a nil pointer possible. And can it really do any damage in the OAuth code?

Anyway, using `Competition.exists?(params[:id])` is probably the right thing here.

jfly replied:
> And can it really do any damage in the OAuth code?

No idea, but I'd rather not have to think about it!

> Anyway, using Competition.exists?(params[:id]) is probably the right thing here.

That will ultimately be more work though, right? You'll have to explicitly handle the "it does not exist" case, whereas with find, a 404 automatically happens.

larspetrus commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -262,6 +262,64 @@ def edit
+      scope: 'read_write',
+      state: @competition.id,
+    }
+    @authorise_url = client.auth_code.authorize_url(oauth_params)
```
"authorise" is as valid a spelling as "authorize", but I'd go with the Z version, since that's what the function is called.

jfly replied:
Agreed. We're going to switch everything over from british to american spelling at some point, as the WCA is incorporating as a US nonprofit.

jfly commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -262,6 +262,64 @@ def edit
+        { stripe_user_id: @competition.connected_stripe_account_id},
+      )
+      @competition.connected_stripe_account_id = nil
+      @competition.save
```
We should a `save!` here, or check the return value of `.save` to see if it succeeded.

jfly commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -262,6 +262,64 @@ def edit
+    redirect_to competitions_payment_setup_path(competition)
+  end
+
+  private def create_oauth_client
```
Nit: since the WCA website does have a decent amount of code about WCA OAuth, I think it might be good to rename this method to make it clear that it's for stripe oath.

jfly commented on WcaOnRails/app/controllers/registrations_controller.rb:
```diff
@@ -192,6 +192,47 @@ def register
+      registrations = competition.registrations
+      registration = registrations.find_by_user_id(current_user.id)
+    else
+      flash[:danger] = "Please register for the competition first, and then pay the entry fee."
```
I'm confused by this else block. This looks more like the person isn't logged in, rather than not registered for the competition.

jfly commented on WcaOnRails/app/models/registration.rb:
```diff
@@ -73,6 +74,33 @@ def best_solve(event, type)
+  end
+
+  def record_payment(amount, currency_code, stripe_charge_id)
+    registration_payments.create(
```
`create` -> `create!`?

jfly commented on WcaOnRails/app/views/competitions/payment_setup.html.erb:
```diff
@@ -0,0 +1,21 @@
+
+  <% else %>
+    <p>
+      This competition is set up to accepts payments through Stripe
```
"set up to accept", not "set up to accepts"

jfly commented on WcaOnRails/app/views/competitions/payment_setup.html.erb:
```diff
@@ -0,0 +1,21 @@
+  <% else %>
+    <p>
+      This competition is set up to accepts payments through Stripe
+      You can deauthorise this account by clicking the button below
```
nit: deauthorise -> deauthorize.

jfly commented on WcaOnRails/app/views/registrations/edit_registrations.html.erb:
```diff
@@ -33,6 +33,7 @@
             <th class="registration-date" data-sortable="true" data-field="registration-date"><%= t 'registrations.list.registered' %></th>
             <th class="guests"><%= t 'activerecord.attributes.registration.guests' %></th>
             <th class="comments"><%= t 'activerecord.attributes.registration.comments' %></th>
+            <th class="Paid"><%= t 'activerecord.attributes.registration.paid_entry_fees' %></th>
```
nit: slightly surprised to a see a css class name capitalized here

jfly commented on WcaOnRails/app/views/registrations/register.html.erb:
```diff
@@ -66,15 +67,46 @@
+              <% if @competition.connected_stripe_account_id %>
+                <% if @competition.connected_stripe_account_id && @registration.outstanding_entry_fees > 0 %>
+                  <% if @registration.paid_entry_fees > 0 %>
+                    <%= alert :info do %>
```
Not something we have to do as part of this PR, but we're definitely going to want to translate all this later on. Could you either do it now, or make an issue for it so we can get it later?

pingskills commented:
I updated the code based on the feedback, added some support for translations, but left the payment_setup method as is for the moment.

pingskills commented:
Currently when someone makes a payment it is stored in a `registration_payments` table. After a user had paid should we allow them to delete their registration and should we delete the payment?

timhabermaas commented:
Usually you want to keep a perfect historical record when money is involved, but maybe Stripe already takes care of this. Speaking of Stripe vs. our DB: Is there a better way to keep the two databases in sync? (two-phase commit?) As it's currently implemented we would charge the user through Stripe, but wouldn't keep track of it when the server/database crashes right after charging.

Besides that: I don't see any code preventing the user from being charged twice.

timhabermaas commented:
> but if something fails, we don't have a record of it, just an error message for the user.

If recording charging attempts is really important we should record them as separate events and not rely on the async worker system for doing so. So, I think the question comes down to: sync vs. async. in terms of user experience.

timhabermaas commented:
> Also, it would be great to have whatever testing we can have for this. Switching to a jobs based infrastructure might make it easier to test our registration/payment flow without having to actually mock out the Stripe gem, we could do all the mocking out at the jobs level.

I would either:

a) build a small Sinatra web server which mimics a small subset of the Stripe API (e.g. rejecting invalid API keys) and mock it using [Webmock](https://robots.thoughtbot.com/how-to-stub-external-services-in-tests#create-a-fake-hello-sinatra) or start it as a separate process and don't stub anything. Or
b) Create our own (domain specific) wrapper around the Stripe API (that's good practice anyway - don't stub what you don't own, especially with dynamically typed languages) and stub that. You'd also need to test the wrapper against the real API.

ronaldmansveld commented:
Based on my experience (with payment providers in general, not stripe
specific):

- log everything, never rely on the payment provider to have the records
available that you need for your use case (I've had some very bad
experiences with this). Have your own logs on initiated payments, and mark
those as paid based on the server to server call coming from the payment
provider. Deleting records that deal with payments should never be done, as
you always need to be in a position to prove what happened to what payment.
Instead of deleting, you can mark the record as nonlonger needed, or create
a refund record where necessary.
- preferably route everything through your own API. This gives flexibility
in changing/adding payment providers, and makes testing easier. This API
endpoint should return the URL to forward the user to to complete the
payment.
- prevent double payments. Payments should be prevented unless the status
of a payment is 'new' or 'failed' (or similar statuses). Not just already
paid records should be prevented from paying again, also those that are in
progress (as some credit cards may take over a day to come through)

On async vs sync: while async is preferred for a lot of operations, sync is
generally preferred for payments. Users don't mind waiting while their
payment is processed (most actually expect it), plus it gives much better
follow up possibilities when a payment is declined or failing for another
reason. Here in the UK I doubt you even can make it async, as most, if not
all, payment providers will have to redirect the user to their respective
bank (when using debit cards) for verification. Even though manual
verification is rarely needed (never had to myself), it is a possibility,
and making payments an async process makes it very hard to do this.

Just my two cents.

Ronald

On Mon, 14 Nov 2016, 2:01 pm Tim Habermaas, <notifications@github.com>
wrote:

> Also, it would be great to have whatever testing we can have for this.
> Switching to a jobs based infrastructure might make it easier to test our
> registration/payment flow without having to actually mock out the Stripe
> gem, we could do all the mocking out at the jobs level.
>
> I would either:
>
> a) build a small Sinatra web server which mimics a small subset of the
> Stripe API (e.g. rejecting invalid API keys) and mock it using Webmock
> <https://robots.thoughtbot.com/how-to-stub-external-services-in-tests#create-a-fake-hello-sinatra>
> or start it as a separate process and don't stub anything. Or
> b) Create our own (domain specific) wrapper around the Stripe API (that's
> good practice anyway - don't stub what you don't own, especially with
> dynamically typed languages) and stub that. You'd also need to test the
> wrapper against the real API.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/thewca/worldcubeassociation.org/pull/1001#issuecomment-260341905>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATRSVjqlS565fofufRpiTUavkqpFJ4_ks5q-GmSgaJpZM4KtHwW>
> .
>

jfly commented:
Some very useful comments on this thread, thanks all! I'm convinced that doing payments synchonously is the right way to go.

> After a user had paid should we allow them to delete their registration and should we delete the payment? -@pingskills
> Deleting records that deal with payments should never be done, as
you always need to be inDeleting records that deal with payments should never be done, as
you always need to be in a position to prove what happened to what payment.
Instead of deleting, you can mark the record as nonlonger needed, or create
a refund record where necessary. a position to prove what happened to what payment.
Instead of deleting, you can mark the record as nonlonger needed, or create
a refund record where necessary. -@ronaldmansveld

I agree, we should not actually delete registrations. Furthermore, I think it would be good if we instead track every change that happens to a registration (and who did it), so we have an audit log. I've created #1005 for this. @pingskills, how do you feel about taking on #1005 first and then coming back to implement payments on top of it?

pingskills commented:
I've added the ability to refund payments through the edit registrations page.

jfly commented on WcaOnRails/app/views/registrations/register.html.erb:
```diff
@@ -66,15 +67,39 @@
+                      data-name="World Cube Association"
+                      data-description="Registration fee"
+                      data-amount="<%= @registration.outstanding_entry_fees.cents %>"
+                      data-currency="<%= @registration.entry_fee_currency_code %>"
```
Just curious... would we be able to get the currency code from `@registration.outstanding_entry_fees` instead? I feel like always interacting with the Money object would minimize our "api surface" in a nice way.

pingskills replied:
Yes we can :)

jfly commented on WcaOnRails/app/views/registrations/register.html.erb:
```diff
@@ -66,15 +67,39 @@
+                      data-email="<%= @registration.user.email %>"
+                      data-zip-code="false"
+                      data-label="<%= t('registrations.payment_button_text', entry_fee: outstanding_entry_fees) %>"
+                      data-locale="auto">
```
Could you work with @viroulep to figure out what this should be set to? Maybe all we need to do here is make an issue for this for @viroulep to deal with later.

jfly commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -539,6 +544,8 @@ en:
       confirmed_not_visible: "This competition is confirmed but not visible"
       not_confirmed_visible: "This competition is not confirmed but visible"
       not_confirmed_not_visible: "This competition is not confirmed and not visible"
+      stripe_connected: "You have now successfully connected your Stripe account and can now recieve registration payments for this competition."
```
nit: recieve -> receive (I before E except after C!)

jfly commented on WcaOnRails/config/routes.rb:
```diff
@@ -51,6 +53,9 @@
 
   get 'competitions/:id/edit/admin' => 'competitions#admin_edit', as: :admin_edit_competition
+  get 'competitions/:id/payment_setup' => 'competitions#payment_setup', as: :competitions_payment_setup
+  get 'competitions/:id/revoke_stripe_access' => 'competitions#revoke_stripe_access', as: :competitions_revoke_stripe_access
```
Am I correct in my understanding that this is a GET route that has a side effect? If so, this should be a POST.

jfly commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -764,4 +764,86 @@
+      sign_out
+
+      it 'redirects to the sign in page' do
+        get :payment_setup, id: competition, competition: { name: competition.name }
```
What's this `competition: { name: competition.name }` part of the `get ...` all about?

pingskills replied:
Good question. I don't know what I was thinking. I must have copied from another test and left it in. I'll get rid of them.

jfly commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -764,4 +764,86 @@
+      sign_in { FactoryGirl.create :admin }
+
+      it 'displays payment setup status' do
+        get :payment_setup, id: competition, competition: { name: competition.name }
```
What's this `competition: { name: competition.name }` part of the `get ...` all about?

jfly commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -764,4 +764,86 @@
+
+      it 'does not allow access' do
+        expect {
+          get :payment_setup, id: competition, competition: { name: competition.name }
```
What's this `competition: { name: competition.name }` part of the `get ...` all about?

jfly commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -764,4 +764,86 @@
+    end
+  end
+
+  describe 'GET #revoke_stripe_access' do
```
Again, this should probably be a POST, right?

larspetrus commented on WcaOnRails/app/controllers/registrations_controller.rb:
```diff
@@ -196,6 +196,70 @@ def register
+    token = params[:stripeToken]
+    Stripe.api_key = ENVied.STRIPE_API_KEY
+
+    begin
```
I think it's better to have the `rescue` clauses at the end of the function and not have the `begin/end` block. This removes some code clutter, separates error handling from the regular flow, and makes you not have to remember to put a `return` at the end of each rescue.

Same goes for the one in `refund_payment`

pingskills commented on WcaOnRails/app/controllers/registrations_controller.rb:
```diff
@@ -196,6 +196,70 @@ def register
+
+    begin
+      charge = Stripe::Charge.create({
+        amount: registration.outstanding_entry_fees.cents,
```
hmm, in the normal application flow yes. If someone can post to this URL without `current_user` set then no.

pingskills commented on WcaOnRails/app/models/registration.rb:
```diff
@@ -101,6 +102,45 @@ def best_solve(event, type)
+  end
+
+  def outstanding_entry_fees
+    entry_fee.nil? ? nil : entry_fee - Money.new(paid_entry_fees, entry_fee_currency_code)
```
Good pickup.

pingskills commented:
Update the code based on some feedback.

Used Money object for getting the currency code in Stripe calls.
Moved rescue blocks to the end of the method.
Made revoke_stripe_access a POST request.

@jfly you also mentioned Stripe testing and 2 phase commits. I agree the testing is extremely beneficial. With the 2 phase commits I'm not sure it adds a lot of value. The worst case scenario as it stands now, is that a charge to Stripe succeeds but we don't record that in our database. We can still confirm the payment via the Stripe account. If we had written a pre-charge row to the database we are in pretty much the same situation. We know we tried to make a charge but it didn't succeed, so we need to check Stripe to see if the charge was successful. By using 2 phase commits though we could have a background process checking for any payments that we attempted but didn't record and then email the organisers?

I think that the chances of us creating a payment and not writing it to the database are extremely small, and if it happens we can still verify it through Stripe. Just my thoughts and I am more than happy to add an extra call to the database beforehand if other people think it is worthwhile.

larspetrus commented on WcaOnRails/app/controllers/registrations_controller.rb:
```diff
@@ -196,6 +196,64 @@ def register
+  rescue => e
+    flash[:danger] = 'Something went wrong with the refund: ' + e.message
+    redirect_to edit_registration_path(registration)
+    return
```
Line 226, 230, 254:
`return` statements aren't needed in method scope rescues.

jfly commented:
> The worst case scenario as it stands now, is that a charge to Stripe succeeds but we don't record that in our database.

I'm sold! Let's go without 2 phase commits for now, we can always add it later if we run into weird issues.

pingskills commented:
I added some integrations tests which connect with Stripe. They failed with the following message: 

_Stripe no longer supports API requests made with TLS 1.1. Please initiate HTTPS connections with TLS 1.2 or later. You can learn more about this at https://stripe.com/blog/upgrading-tls._

What do people think about an Integration test like this in the test suite? And if it's OK, is there a way to upgrade Travis to TLS1.2?

jfly commented on WcaOnRails/Envfile:
```diff
@@ -31,3 +31,7 @@ variable :RECAPTCHA_PRIVATE_KEY, :string, default: ''
+variable :STRIPE_API_KEY, :string, default: 'sk_test_vEAI1zKp934C2ML0BUPHziXd'
+variable :STRIPE_PUBLISHABLE_KEY, :string, default: 'pk_test_A1Ne6QlSRbKvnL24t3gbTva0'
+variable :STRIPE_CLIENT_ID, :string, default: 'ca_9SMJGf8Xc4MhDNdmEWMRJuWBraH3tNRk'
+variable :STRIPE_REDIRECT_URI, :string, default: 'http://localhost:3000/stripe-connect'
```
Is this `STRIPE_REDIRECT_URI` env var used anywhere?

pingskills replied:
It doesn't appear so. I'll get rid of it. This is actually set in the Stripe website.

jfly commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -272,6 +272,64 @@ def edit
+  def revoke_stripe_access
+    @competition = Competition.find(params[:id])
+    if @competition.connected_stripe_account_id
+      Stripe.api_key = ENVied.STRIPE_API_KEY
```
Bump. Any word on this?

jfly commented on WcaOnRails/app/controllers/registrations_controller.rb:
```diff
@@ -196,6 +196,61 @@ def register
+      registration = registrations.find_by_user_id!(current_user.id)
+    end
+    token = params[:stripeToken]
+    Stripe.api_key = ENVied.STRIPE_API_KEY
```
I'm still kind of weirded out by this. Would this be more appropriate as a rails intializer? See https://github.com/jfly/worldcubeassociation.org/blob/master/WcaOnRails/config/initializers/delayed_job_config.rb for an example. Docs here http://guides.rubyonrails.org/configuring.html#using-initializer-files.

pingskills replied:
Thanks for the links. I'll put it into a stripe configuration file.

jfly commented on WcaOnRails/spec/controllers/registrations_controller_spec.rb:
```diff
@@ -562,4 +562,105 @@
+      end
+
+      it 'processes payment with valid credit card' do
+        Stripe.api_key = ENVied.STRIPE_API_KEY
```
Same comment as above about if this really belongs here.

jfly commented on WcaOnRails/spec/controllers/registrations_controller_spec.rb:
```diff
@@ -562,4 +562,105 @@
+      end
+
+      it 'rejects payment with invalid credit card' do
+        Stripe.api_key = ENVied.STRIPE_API_KEY
```
Same comment as above about if this really belongs here.

jfly commented on WcaOnRails/spec/controllers/registrations_controller_spec.rb:
```diff
@@ -562,4 +562,105 @@
+
+      it 'processes refund' do
+        sign_in organizer
+        Stripe.api_key = ENVied.STRIPE_API_KEY
```
Same comment as above about if this really belongs here.

jfly commented on WcaOnRails/spec/controllers/registrations_controller_spec.rb:
```diff
@@ -562,4 +562,105 @@
+        ).id
+        post :process_payment, competition_id: competition.id, stripeToken: token_id
+        expect(response).to redirect_to competition_register_path(competition)
+        expect(registration.reload.outstanding_entry_fees).to eq 0
```
It's not super clear to me reading this test that `outstanding_entry_fees` actually changed here. Maybe it just started out as 0? To make this clearer, I'd vote for adding a `expect` before the POST that proves that `outstanding_entry_fees` is greater than zero before the payment.

jfly commented on WcaOnRails/spec/controllers/registrations_controller_spec.rb:
```diff
@@ -562,4 +562,105 @@
+          },
+        ).id
+        post :process_payment, competition_id: competition.id, stripeToken: token_id
+        expect(response).to redirect_to competition_register_path(competition)
```
Would it make sense to check for a flash message here as well?

pingskills replied:
I probably should have a flash success message here.

jfly commented on WcaOnRails/spec/controllers/registrations_controller_spec.rb:
```diff
@@ -562,4 +562,105 @@
+      let!(:user) { FactoryGirl.create(:user, :wca_id) }
+      let!(:registration) { FactoryGirl.create(:registration, competition: competition, user: user) }
+
+      it 'redirects to the sign in page' do
```
This description doesn't seem to match what the test is actually checking for.

jfly commented:
This looks relevant for stripe on Ubuntu 12.04, which is what travis is using (see [here](https://docs.travis-ci.com/user/ci-environment/)): https://support.stripe.com/questions/how-do-i-upgrade-my-openssl-to-support-tls-1-2. I'd try adding the recommended openssl upgrade commands to our travis config https://github.com/thewca/worldcubeassociation.org/blob/f0edca737dc691370d4e2f48e64ac57ed0f5b66d/.travis.yml#L11 and see what happens. Good luck!

pingskills commented:
As suggested by @jfly I've created a stripe initialization file, and updated the stripe specific tests.

jfly commented:
This is starting to look great! How do you want to test this out? I'd be okay merging this to production if that's safest (staging isn't exactly a secure environment). 

As discussed over video chat, we need to make a wca stripe account for this to work, right? Can you look into doing that? Don't use any of your personally identifying information to do it if you can avoid it, and put the username and password in the secrets doc. After that, we'll have to set this up with our chef databags, which we should have instructions for in our readme.

pingskills commented:
I'd prefer to deploy this to staging first and have someone else help test the process. We'll be exposing test keys but I feel that is OK. 

I'll create a WCA Stripe account.

pingskills commented:
@jfly - I've tried updating the chef data_bag with the new production stripe keys, and modified the template to use these values for production, and use the Envfile for development, and staging. When you have some time can you please take a look and see if I'm on the right track with these changes? Thanks!

jfly commented:
Your latest commit looks very similar to what I had in mind, go for it!

I thought @viroulep had gotten rid of the -noregs stuff, but I guess that's still floating around.

viroulep commented on chef/site-cookbooks/wca/templates/env.production.erb:
```diff
@@ -1,7 +1,7 @@
 WCA_LIVE_SITE=<%= node.chef_environment == "production" ? "1" : "0" %>
 ROOT_URL=<%=
   {
-    "development" => "", "development-noregs" => "", # Envfile will compute a sane default for us
```
@jfly indeed this one slept through my search...
Thanks @pingskills!

pingskills commented:
I updated the call to Stripe when connecting accounts to pass along a redirect_uri. In the Stripe account you specify a comma separated list of accepted uris. This allows the code to work locally, and on staging.

I also updated the competitions factory to use the newly created WCA Test Organizer account. This is so our tests will call the real Stripe API using this connected account.

jfly commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -276,6 +276,64 @@ def edit
+    client = create_stripe_oauth_client
+    oauth_params = {
+      scope: 'read_write',
+      redirect_uri: ENVied.ROOT_URL + '/stripe-connect',
```
Could we instead call a Rails URL helper here? Something like `competitions_stripe_connect_url`? I think all this would require is adding a `as: :competitions_stripe_connect` to the corresponding route in our routes file.

jfly commented on WcaOnRails/config/routes.rb:
```diff
@@ -51,6 +53,9 @@
   get 'competitions/:id/edit/admin' => 'competitions#admin_edit', as: :admin_edit_competition
+  get 'competitions/:id/payment_setup' => 'competitions#payment_setup', as: :competitions_payment_setup
+  post 'competitions/:id/revoke_stripe_access' => 'competitions#revoke_stripe_access', as: :competitions_revoke_stripe_access
+  get 'stripe-connect' => 'competitions#stripe_connect'
```
This might need to get a `as: :competitions_stripe_connect` for my other suggestion to work. You should run `bin/rake routes` first to verify that there isn't already a "Prefix" for this.

jfly commented on WcaOnRails/spec/factories/competitions.rb:
```diff
@@ -76,5 +76,11 @@
+    trait :entry_fee do
+      base_entry_fee_lowest_denomination 1000
+      currency_code "AUD"
+      connected_stripe_account_id "acct_19ZQVmE2qoiROdto"
```
Your call if you think it's valuable, but I might be inclined to comment on where this id came from, and how to change/recreate it. I'm probably going to forget that we actually have 2 stripe accounts.

A side note: what will happen in production if someone connects a Stripe account in in testing mode? Just their mistake, and people will be able to register without actually paying anything?

pingskills replied:
@jfly - I emailed Stripe support and they inform me that if the WCA uses their live keys, then it will only connect to a live account so a user cannot connect with their test keys to our live keys.

From Stripe support:

> SInce the platform is using the Live API Key, the platform will treat any information passed along the API as live data. If test details are being entered through your onboarding process for the platform, your user's should see an error saying that they are entering known test details. Ultimately, in this scenario a account couldn't connect with this information and the Dashboard being in test/live mode won't have an effect.

jfly commented on .travis.yml:
```diff
@@ -23,7 +24,7 @@ before_install:
 before_script:
   - bundle exec rake db:reset RAILS_ENV=test
-sudo: false
+sudo: required
```
It's a shame that we'll have to lose Travis's container-based infrastructure here (see [this doc](https://docs.travis-ci.com/user/migrating-from-legacy/#Why-migrate-to-container-based-infrastructure%3F) explaining why it's nicer than their legacy vm infrastructure).

It looks like it's possible for us to opt-in to Ubuntu 14.04 (Trusty): https://docs.travis-ci.com/user/trusty-ci-environment/.  That version of Ubuntu might have a new enough version of openssl installed that things will Just Work, and would let us stick with the container environment. Could you give that a shot?

pingskills replied:
@jfly - it looks like MySQL has been [removed from trusty](https://github.com/travis-ci/docs-travis-ci-com/commit/d71108826c0de8387ce57c5c0519c7bce08e5cac). It looks like this means we'd need to use `sudo` to install it and hence couldn't use the container based infrastructure anyhow.

jfly replied:
Damn! Can't win =( Sounds like we'll have to go containerless, and hope that Travis gets their act together.

jfly replied:
Strange, I guess this documentation is out of date: https://docs.travis-ci.com/user/database-setup/#MySQL-5.6. It says "The recommended way to get MySQL 5.6 is switching to our Trusty CI Environment."

pingskills replied:
It is strange. I thought MySQL would be extremely popular. But on the Trusty CI environment, if you look at data stores, MySQL is missing. https://docs.travis-ci.com/user/trusty-ci-environment

I'm going away for a week but when I get back I'll try to get the PR finished up so we can merge it up and then move on to fixing the registration flow.

pingskills commented:
The Travis environment has been reset to the original with a sudo call to upgrade openssl. I've added some comments as suggested, and removed the link in the admin section to the payment setup. If a competition doesn't set up payments then the new registration code won't be used.

pingskills commented:
Rebased the PR to merge in other recent changes. Let me know your thoughts on merging this up, or if there are still some concerns you want addressed first.

jfly commented on WcaOnRails/app/views/competitions/_nav.html.erb:
```diff
@@ -35,7 +35,7 @@
           children: [
             { text: t('.menu.orga_view'), path: edit_competition_path(@competition), fa_icon: "lock" },
-            { text: t('.menu.event_view'), path: edit_events_path(@competition), fa_icon: "lock" },
+            { text: t('.menu.event_view'), path: edit_events_path(@competition), fa_icon: "cubes" },
```
Good call! I found those two locks very confusing.

jfly commented on WcaOnRails/app/views/registrations/edit.html.erb:
```diff
@@ -30,20 +30,39 @@
+  <% end %>
+  <hr />
+  <div>
+    <h4>Payment History</h4>
```
Hmm... would it make sense to only show this if the competition actually has payments turned on?

pingskills commented:
Rebased to merge in the latest changes. Updated a few small items based on feedback from @jfly

pingskills commented:
Good question @jfly - I just set the values back to what they were in my file, and kept the addition of the GITHUB_BOT_ACCESS_TOKEN. So I'm not sure if this works. Do I need to generate the entire file again on staging, and then put that in source control?

jfly commented:
Yeah, I also am unsure how this works. I can say from experience that the entire file seems to change whenever you change one thing, so I suspect your actions may have corrupted the `GITHUB_BOT_ACCESS_TOKEN`? I'd try running knife show on the current version in `master` vs your changed version and see what happens. I'm very curious to see what happens here.

pingskills commented:
Good pickup @jfly - I get a different value for GITHUB_BOT_ACCESS_TOKEN. Looks like I need to edit the file on staging and then copy that back to my repository.

jfly commented:
:ship: :shipit: !!!

Too bad about the databags, those are definitely annoying to work with.

pingskills commented:
Hi Jeremy,

It seems unicorn has died again :) I'm looking into it now.

Regards,
Jeff.


On 1 February 2017 at 14:56, Jeremy Fleischman <notifications@github.com>
wrote:

> 🚢 [image: :shipit:] !!!
>
> Too bad about the databags, those are definitely annoying to work with.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/thewca/worldcubeassociation.org/pull/1001#issuecomment-276566357>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMv4JlsOh83XPG3DACddf-p2ZyOW0Opks5rYAJzgaJpZM4KtHwW>
> .
>

jfly commented:
Looks good now! I should have reminded you about https://github.com/thewca/worldcubeassociation.org/issues/1203 =)

pingskills commented:
All fixed after running chef-solo and restarting unicorn.

Regards,
Jeff.

Table Tennis Secrets Revealed
http://www.PingSkills.com

On 1 February 2017 at 15:11, Jeff Plumb <jwplumb@gmail.com> wrote:

> Hi Jeremy,
>
> It seems unicorn has died again :) I'm looking into it now.
>
> Regards,
> Jeff.
>
>
> On 1 February 2017 at 14:56, Jeremy Fleischman <notifications@github.com>
> wrote:
>
>> 🚢 [image: :shipit:] !!!
>>
>> Too bad about the databags, those are definitely annoying to work with.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/thewca/worldcubeassociation.org/pull/1001#issuecomment-276566357>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AEMv4JlsOh83XPG3DACddf-p2ZyOW0Opks5rYAJzgaJpZM4KtHwW>
>> .
>>
>
>
